tsl/src/chunk_copy.c

```diff
@@ -547,33 +698,66 @@ chunk_copy_stage_delete_chunk(ChunkCopy *cc)
 }
 
 static const ChunkCopyStage chunk_copy_stages[] = {
+	/* Initial Marker */
+	{ CCS_INIT, NULL, chunk_copy_stage_init_cleanup },
 
-	/* Create empty chunk table on the dst node */
-	{ CCS_CREATE_EMPTY_CHUNK, chunk_copy_stage_create_empty_chunk },
+	/*
+	 * Create empty chunk table on the dst node.
+	 * The corresponding cleanup function should just delete this empty chunk.
+	 */
+	{ CCS_CREATE_EMPTY_CHUNK,
+	  chunk_copy_stage_create_empty_chunk,
+	  chunk_copy_stage_create_empty_chunk_cleanup },
 
-	/* Setup logical replication between nodes */
-	{ CCS_CREATE_PUBLICATION, chunk_copy_stage_create_publication },
-	{ CCS_CREATE_REPLICATION_SLOT, chunk_copy_stage_create_replication_slot },
-	{ CCS_CREATE_SUBSCRIPTION, chunk_copy_stage_create_subscription },
+	/*
+	 * Setup logical replication between nodes.
+	 * The corresponding cleanup functions should drop the subscription and
+	 * remove the replication slot followed by dropping of the publication on
+	 * the source data node.
+	 */
+	{ CCS_CREATE_PUBLICATION,
+	  chunk_copy_stage_create_publication,
+	  chunk_copy_stage_create_publication_cleanup },
+	{ CCS_CREATE_REPLICATION_SLOT,
+	  chunk_copy_stage_create_replication_slot,
+	  chunk_copy_stage_create_replication_slot_cleanup },
+	{ CCS_CREATE_SUBSCRIPTION,
+	  chunk_copy_stage_create_subscription,
+	  chunk_copy_stage_create_subscription_cleanup },
 
-	/* Begin data transfer and wait for completion */
-	{ CCS_SYNC_START, chunk_copy_stage_sync_start },
-	{ CCS_SYNC, chunk_copy_stage_sync },
+	/*
+	 * Begin data transfer and wait for completion.
+	 * The corresponding cleanup function should just disable the subscription so
+	 * that earlier steps above can drop the subcription/publication cleanly.
+	 */
+	{ CCS_SYNC_START, chunk_copy_stage_sync_start, chunk_copy_stage_sync_start_cleanup },
+	{ CCS_SYNC, chunk_copy_stage_sync, NULL },
 
-	/* Cleanup */
-	{ CCS_DROP_PUBLICATION, chunk_copy_stage_drop_publication },
-	{ CCS_DROP_SUBSCRIPTION, chunk_copy_stage_drop_subscription },
+	/*
+	 * Cleanup. Nothing else required via the cleanup functions.
+	 */
+	{ CCS_DROP_SUBSCRIPTION, chunk_copy_stage_drop_subscription, NULL },
+	{ CCS_DROP_PUBLICATION, chunk_copy_stage_drop_publication, NULL },
 
-	/* Attach chunk to the hypertable on the dst_node */
-	{ CCS_ATTACH_CHUNK, chunk_copy_stage_attach_chunk },
+	/*
+	 * Attach chunk to the hypertable on the dst_node.
+	 * The operation has succeeded from the destination data node perspective.
+	 * No cleanup required here.
+	 */
+	{ CCS_ATTACH_CHUNK, chunk_copy_stage_attach_chunk, NULL },
```

Comment:
We don't need clean for it, because it is being run in the transaction?

Reply:
The cleanup will actually happen when the "DROP TABLE" and metadata cleanup happens in other cleanup function.